features/323-login-id/design.md

```diff
@@ -0,0 +1,253 @@
+**NOTE: this specification deprecates #294: Login without specifiying login ID key**
+
+# Background
+
+## Current Design
+
+Under current design of authentication:
+
+1. Each **User** is associated with some **Principal**.
+3. Each **Principal** is associated with one **Identity Provider** (e.g. password authentication, OAuth).
+4. Password authentication principals is represented by tuple (**Login ID Key**, **Login ID**, password hash).
+5. App developer specifies a list of white-listed login ID key.
+6. To signup using password authentication, user can provide a list of tuple (login ID key, login ID). e.g. `{ "username": "test", "email": "a@example.com" }`
+7. For password authentication principals, user can login using a tuple (login ID key, login ID). e.g. `{ "email": "a@example.com" }`
+8. Login IDs are globally unique in app, without considering associated login ID keys.
+9. Auth gear does not knows the semantics of login IDs.
+
+
+## Problem
+1. Some features requires semantics of login IDs to function.
+2. Some authentication system designs may not be implemented easily in current design.
+
+
+# Use cases
+
+Considering some practical scenarios of authentication system:
+
+## GitHub
+1. Supports multiple emails per user, with email verification.
+2. User can add/delete any amount of emails.
+3. Only verified emails would be associated with the user in public.
+4. User can allow recovery email/security alert to be sent to all verified emails or a specific verified email.
+
+
+## Facebook
+5. User can login with email/phone number/username, without indicating what type of value it is.
+6. User can add/delete any amount of emails, which can be used to login after verification.
+7. Disposable email address is blocked.
+8. All emails/phones of user is verified.
+9. Email communications are personalized with user language & name.
+
+## Other systems
+10. User may choose to login as different roles (e.g. teacher/student, admin/member).
+
+
+# Design Objectives
+1. Allow auth related features to understand semantics of some login IDs
+2. Support multiple login ID with same login ID key
+3. Support compound login IDs
+4. Support adding/removing login ID
+
+
+# Proposed Design Overview
+
+## Use Implicit Login ID Key for Logging in
+In current design, user would login using tuple (login ID key, login ID). i.e:
```

Comment:
again.. this is wrong, the spec allow login using login ID only

Reply:
The spec that allows login using only login ID is #294, which is deprecated by this document. Also, that spec had not yet implemented. Therefore I refer the design implemented as 'current design', and #294 as 'previous design'.

I suppose the terms I used is confusing. I'll change 'current design' to 'previous design', and 'previous design' to 'deprecated design'.

Reply:
Just avoid using terms or sentences that don't convey any useful information.